pulsar-client/src/main/java/org/apache/pulsar/client/impl/conf/ClientConfigurationData.java

```diff
@@ -367,16 +370,49 @@ public class ClientConfigurationData implements Serializable, Cloneable {
     @Secret
     private String socks5ProxyPassword;
 
+    /**
+     * Gets the authentication settings for the client.
+     * <p>
+     * If the authentication has not been specified an attempt is made to determine one using auth-related
+     * properties. In this case, the {@code authPluginClassName} and one of {@code authParams} or {@code authParamMap}
+     * must be configured for the authentication to be determined. Otherwise, a default of
+     * {@link AuthenticationDisabled} is returned.
+     *
+     * @return authentication settings for the client
+     */
     public Authentication getAuthentication() {
         if (authentication == null) {
-            this.authentication = AuthenticationDisabled.INSTANCE;
+            setAuthenticationFromPropsIfAvailable();
```

Comment:
usually it is not a good practice to mutate the state in a "getter".

ClientConfigurationData is a POJO that defines the configuration, it shouldn't perform this kind of operations.

I suggest to move this logic to a new utility method and to not change the getter (that already does some automatic fix up, that we should remove, and simply return AuthenticationDisabled.INSTANCE without updating the field).

Also updating the fields introduces concurrency issues, I think that we shouldn't update the POJO

Reply:
I think we got to this point out of convenience. A few commits back it was doing the utility method approach and that utility was being called in the proper places. I agree w/ the concerns and am happy to move this back to https://github.com/apache/pulsar/pull/18358/commits/eeb182c092f215d2741fa18e48c372450c464752 . @nodece wdyt?

Reply:
It's all fine with me. Right now implementation is just for convenience, and keep having one method to create authentication to avoid duplication coding.

Reply:
I agree with @eolivelli that a getter should not mutate state. Maybe the logic can be moved to the setters ?

Reply:
Thanks @cbornet I will move it back something like it was a few commits ago and/or explore the setters idea. We can all re-review at that time and see if we find some consensus. 😄